File: packages/cli/src/commands/redeem.ts
import type { Command } from 'commander';
import { parseVaa, WORMHOLESCAN_MAINNET, WORMHOLESCAN_TESTNET } from '@worm-tool/sdk';
import { loadConfig } from '../config.js';
import { createEvmChain } from '../providers/evm.js';
import { printJson, printError } from '../output.js';
import { abiEncodeBytes } from './_vaa-utils.js';

function isTxHash(input: string): boolean {
  return input.startsWith('0x') && input.length === 66 && /^[0-9a-fA-F]+$/.test(input.slice(2));
}

export function registerRedeemCommand(program: Command): void {
  program
    .command('redeem <input>')
    .description('Manually redeem a Wormhole VAA on the destination EVM chain (input: tx hash or raw VAA hex/base64)')
    .requiredOption('--chain <name>', 'Destination EVM chain name (e.g. ethereum)')
    .option('--contract <address>', 'Target contract address')
    .option('--selector <hex>', 'Function selector override (default: completeTransfer 0xc6878519)')
    .option('--network <network>', 'mainnet or testnet (used if input is a tx hash)', 'mainnet')
    .action(async (input: string, opts: { chain: string; contract?: string; selector?: string; network: string }) => {
      try {
        const config = loadConfig();
        const network = opts.network === 'testnet' ? 'testnet' : 'mainnet';

        let vaaHex: string;
        if (isTxHash(input)) {
          const base = network === 'testnet' ? WORMHOLESCAN_TESTNET : WORMHOLESCAN_MAINNET;
          const res = await fetch(`${base}/api/v1/transactions/${input}`);
          if (!res.ok) throw new Error(`Failed to fetch VAA for tx ${input}: ${res.status}`);
          const data = await res.json() as { data?: { vaa?: { raw?: string } } };
          const raw = data.data?.vaa?.raw;
          if (!raw) throw new Error(`No VAA found for tx ${input} (not yet signed?)`);
          vaaHex = raw.startsWith('0x') ? raw : '0x' + raw;
        } else {
          parseVaa(input);
          vaaHex = input.startsWith('0x') ? input : '0x' + input;
        }

        const chain = createEvmChain(opts.chain, config);
        if (!opts.contract) throw new Error(`--contract is required for ${opts.chain}`);

        const selectorRaw = opts.selector ?? '0xc6878519';
        const selector = selectorRaw.startsWith('0x') ? selectorRaw : '0x' + selectorRaw;
        const data = (selector + abiEncodeBytes(vaaHex).slice(2)) as `0x${string}`;
        const receipt = await chain.sendTransaction(opts.contract, data);
        printJson({ txHash: receipt.txHash, success: receipt.success, chain: opts.chain });
      } catch (err) {
        printError('redeem failed', err);
        process.exit(1);
      }
    });
}

File: archive/rust-crates/wormhole-cli/src/commands/redeem.rs
use clap::Args;
use wormhole_sdk::{
    chains::{
        evm::{self, SubmitParams, DEVNET_CORE, DEVNET_FROM, DEVNET_RPC},
        Network,
    },
    status,
    vaa::decode_vaa_bytes,
};

use crate::output;

/// Manually redeem a stuck Wormhole VAA on the destination chain.
#[derive(Debug, Args)]
pub struct RedeemArgs {
    /// Source tx hash (0x + 64 hex chars) or a raw VAA (hex or base64).
    pub input: String,
    /// Destination chain name (e.g. ethereum).
    #[arg(long)]
    pub dst_chain: Option<String>,
    /// Destination chain RPC endpoint.
    #[arg(long, default_value = DEVNET_RPC)]
    pub rpc: String,
    /// `from` address for devnet (unlocked account).
    #[arg(long, env = "WORMHOLE_EVM_FROM")]
    pub from: Option<String>,
    /// Target contract address (defaults to devnet core bridge).
    #[arg(long)]
    pub contract: Option<String>,
    /// Wormhole network for API lookup when input is a tx hash.
    #[arg(long, default_value = "mainnet")]
    pub network: String,
}

/// Run the `worm redeem` command.
///
/// Accepts either a source transaction hash (fetches the signed VAA from the
/// Wormhole Scan API) or raw VAA bytes (hex or base64).  Submits the VAA to
/// the destination chain and prints the resulting transaction hash.
///
/// # Errors
///
/// Returns an error if the input is invalid, the API lookup fails, or the
/// on-chain submission fails.
pub async fn run(args: &RedeemArgs) -> anyhow::Result<()> {
    let vaa_bytes = if is_tx_hash(&args.input) {
        let network: Network = args
            .network
            .parse()
            .map_err(|e| anyhow::anyhow!("invalid network: {e}"))?;
        status::fetch_vaa_bytes(&args.input, network)
            .await
            .map_err(|e| anyhow::anyhow!("failed to fetch VAA: {e}"))?
    } else {
        decode_vaa_bytes(&args.input).map_err(|e| anyhow::anyhow!("invalid VAA: {e}"))?
    };

    let vaa_hex = hex::encode(&vaa_bytes);
    let contract = args.contract.as_deref().unwrap_or(DEVNET_CORE).to_string();
    let from = args.from.as_deref().or(Some(DEVNET_FROM));

    let tx_hash = evm::submit_vaa(SubmitParams {
        vaa_hex: &vaa_hex,
        rpc_url: &args.rpc,
        contract_address: &contract,
        from_address: from,
        evm_key: None,
    })
    .await
    .map_err(|e| anyhow::anyhow!("redeem failed: {e}"))?;

    output::print_json(&tx_hash)
}

/// Return `true` if `input` looks like an EVM transaction hash.
///
/// A tx hash is `0x`-prefixed followed by exactly 64 lowercase hex characters
/// (total 66 chars).
fn is_tx_hash(input: &str) -> bool {
    input.starts_with("0x")
        && input.len() == 66
        && input[2..].chars().all(|c| c.is_ascii_hexdigit())
}

// ── tests ─────────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn detects_tx_hash() {
        let hash = format!("0x{}", "a".repeat(64));
        assert!(is_tx_hash(&hash));
    }

    #[test]
    fn rejects_non_tx_hash_input() {
        // base64 VAA won't match the tx hash pattern
        assert!(!is_tx_hash("AQAAAAQNAB..."));
        // hex without 0x prefix is a VAA, not a tx hash
        assert!(!is_tx_hash(&"a".repeat(64)));
        // too short
        assert!(!is_tx_hash("0xdeadbeef"));
    }

    #[test]
    fn vaa_api_url_mainnet_has_expected_format() {
        use wormhole_sdk::chains::Network;
        use wormhole_sdk::status::vaa_api_url;
        let url = vaa_api_url(
            2,
            "0000000000000000000000003ee18b2214aff97000d974cf647e7c347e8fa585",
            99,
            Network::Mainnet,
        );
        assert_eq!(
            url,
            "https://api.wormholescan.io/api/v1/vaas/2/0000000000000000000000003ee18b2214aff97000d974cf647e7c347e8fa585/99"
        );
    }
}
